gpflow/monitor.py

```diff
@@ -0,0 +1,245 @@
+# Copyright 2020 GPflow authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Provides basic functionality to monitor optimisation runs """
+
+
+from abc import ABC, abstractmethod
+from io import BytesIO
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import numpy as np
+import tensorflow as tf
+from matplotlib.backends.backend_agg import FigureCanvasAgg
+from matplotlib.figure import Axes, Figure
+
+from .models import BayesianModel
+from .utilities import parameter_dict
+
+
+class MonitorTask(ABC):
+    """
+    A base class for a monitoring task.
+
+    All monitoring tasks are callable objects.
+    A descendant class must implement the `run` method, which is the body of the monitoring task.
+    """
+
+    def __init__(self, period: int = 1):
```

Comment:
nit: I think `interval` would make slightly more intuitive sense than `period` - would you be up for a search&replace on that ?